trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,181 @@
+            ``datasets``.
+        :param fit_model: If `True`, models passed in will be optimized on the given data.
+            If `False`, the models are assumed to be optimized already.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step.
+        :param acquisition_state: The optional acquisition state for stateful acquisitions.
+        :raise ValueError: If any of the following are true:
+            - the keys in ``datasets`` and ``model_specs`` do not match
+            - ``datasets`` or ``model_specs`` are empty
+        """
+        self._search_space = search_space
+        self._acquisition_rule = acquisition_rule
+        self._acquisition_state = acquisition_state
+
+        if not datasets or not model_specs:
+            raise ValueError("dicts of datasets and model_specs must be populated.")
+
+        if isinstance(datasets, Dataset):
+            datasets = {OBJECTIVE: datasets}
+            # ignore below is due to MyPy not being able to handle overloads properly
+            model_specs = {OBJECTIVE: model_specs}  # type: ignore
+
+        # reassure the type checker that everything is tagged
+        datasets = cast(Dict[str, Dataset], datasets)
+        model_specs = cast(Dict[str, ModelSpec], model_specs)
+
+        if datasets.keys() != model_specs.keys():
+            raise ValueError(
+                f"datasets and model_specs should contain the same keys. Got {datasets.keys()} and"
+                f" {model_specs.keys()} respectively."
+            )
+
+        self._datasets = datasets
+        self._models = map_values(create_model, model_specs)
+
+        if fit_model:
+            # there is no default value for `fit_model` parameter
+            # to force the user to think about it
+            # before invoking ask/tell API
+            # there is no sensible default here:
+            # `False` may result in suboptimal performance
+            # `True` may result in two model training calls per optimization step
+            for tag, model in self._models.items():
+                dataset = datasets[tag]
+                model.update(dataset)
+                model.optimize(dataset)
+
+    @classmethod
+    def from_record(
+        cls, search_space: SP, acquisition_rule: AcquisitionRule[S, SP], record: Record[S]
+    ) -> AskTellOptimizer[S, SP]:
+        """Creates new :class:`~AskTellOptimizer` instance from provided optimization state.
+
+        :param search_space: The space over which to search.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step.
+        :param record: Optimization state record.
+        """
+        # here we are recovering previously saved optimization state
+        # there the model was already trained
+        # thus there is no need to train it again
+        fit_model = False
+        return cls(
```

Comment:
With this many arguments, it's probably clearer to use named arguments either for all of them or all but the important first three.
```suggestion
        return cls(
            search_space,
            record.datasets,
            record.models,
            fit_model=False,  # model is already trained
            acquisition_rule=acquisition_rule,
            acquisition_state=record.acquisition_state,
        )        
```

Reply:
agreed